controlplane/kubeadm/controllers/helpers.go

```diff
@@ -46,26 +47,45 @@ func (r *KubeadmControlPlaneReconciler) reconcileKubeconfig(ctx context.Context,
 		return nil
 	}
 
-	_, err := secret.GetFromNamespacedName(ctx, r.Client, clusterName, secret.Kubeconfig)
+	controllerOwnerRef := *metav1.NewControllerRef(kcp, controlplanev1.GroupVersion.WithKind("KubeadmControlPlane"))
+	configSecret, err := secret.GetFromNamespacedName(ctx, r.Client, clusterName, secret.Kubeconfig)
 	switch {
 	case apierrors.IsNotFound(err):
 		createErr := kubeconfig.CreateSecretWithOwner(
 			ctx,
 			r.Client,
 			clusterName,
 			endpoint.String(),
-			*metav1.NewControllerRef(kcp, controlplanev1.GroupVersion.WithKind("KubeadmControlPlane")),
+			controllerOwnerRef,
 		)
 		if createErr != nil {
-			if createErr == kubeconfig.ErrDependentCertificateNotFound {
+			if errors.Is(createErr, kubeconfig.ErrDependentCertificateNotFound) {
 				return errors.Wrapf(&capierrors.RequeueAfterError{RequeueAfter: dependentCertRequeueAfter},
-					"could not find secret %q for Cluster %q in namespace %q, requeuing",
-					secret.ClusterCA, clusterName.Name, clusterName.Namespace)
+					"could not find secret %q, requeuing", secret.ClusterCA)
 			}
 			return createErr
 		}
+		// skip rotation checks
+		return nil
 	case err != nil:
-		return errors.Wrapf(err, "failed to retrieve kubeconfig Secret for Cluster %q in namespace %q", clusterName.Name, clusterName.Namespace)
+		return errors.Wrap(err, "failed to retrieve kubeconfig Secret")
+	}
+
+	// only do rotation on owned secrets
+	if !util.HasOwnerRef(configSecret.OwnerReferences, controllerOwnerRef) {
+		return nil
+	}
+
+	needsRotation, err := kubeconfig.NeedsClientCertRotation(configSecret, certs.ClientCertificateRenewalDuration)
+	if err != nil {
+		return errors.Wrap(err, "failed to decode kubeconfig client certificates")
+	}
+
+	if needsRotation {
+		r.Log.Info("rotating kubeconfig secret")
```

Comment:
```suggestion
```

This should probably be a Condition, e.g. `KubeconfigSecretRotated`, and the timestamp should be updated if the condition is already there

cc @fabriziopandini

Reply:
![condition all the things](https://i.imgflip.com/43z26h.jpg)

Reply:
Do you want to block this feature on #3139 then?

Reply:
I definitely think we should coordinate between this PR and #3139, no real preference from my perspective on order for which they land, though.

Reply:
Yeah I'd say let's either just include it in that PR or file a separate PR for it. This can merge now as is.